### ISSUE TITLE
Getting breathe working

### ISSUE THREAD

**Issue by mrob95** [daanzu/kaldi-active-grammar#3]
Hi, the [dragonfly API I'm working](https://github.com/mrob95/Breathe) on has initial features implemented, but it does a couple of slightly unorthodox things to manage grammars and I'm struggling to debug it on kaldi. I've made an issue to track the bugs [here](https://github.com/mrob95/Breathe/issues/1).

The grammar management works by generating grammars as necessary on the fly during `process_begin`. The implementation is pretty efficient so at least on natlink this is done quickly enough to be imperceptible.

Issues so far:
1. Grammars loaded with empty lists cause an exception during compilation. I am using lists to keep track of contexts which are activated using an "enable" command (as in caster).
2. After all grammars are loaded, the first phrase starts, I get a message saying that a new grammar is being loaded (as expected), and then I just get "press any key to continue..." (implying that the program exited and i'm back in the batch script). It looks like the new grammar is loaded correctly but the recognition callback is never reached. Do you know what might be causing the program to exit? pdb is not catching the exit, so not even sure it's Python code that is causing this but I don't know enough about kaldi to be confident about anything. 
**EDIT: setting lazy_compilation=False seems to fix the problem so I am assuming it is some sort of concurrency bug with grammars being loaded at the same time as recognition occurring.** Initial start-up is still significantly quicker than with a caster-based system, and the slight delay when a new context is encountered is hardly noticeable.

The grammar I'm using to test is just:

```
from dragonfly import *
from breathe import Breathe

Breathe.add_commands(
    None,
    {
        "test": Text("test")
    }
)
```

Thanks!

**Comment by daanzu** [comment]
Thanks for the bug rep, and the simple test! I had considered the case of adding grammars during actions, but not in process_begin. I will take a look.

1. Do you mean ListRefs?
2. Turning on debugging logging is pretty helpful, although I would think Kaldi would print the message if it was a fatal error.

**Comment by mrob95** [comment]
1. yes, DictList is just a dragonfly wrapper for a List which maps its values to some other objects. As far as the engine is concerned they are the same.

2. Yeah from the debug output it looks like the new grammar FST loads correctly, but then almost immediately the program exits with no messages.

**Comment by daanzu** [comment]
@mrob95 Re empty ListRefs: I tried a test grammar with an empty ListRef without Breathe, and it worked fine. Any pointers for a simple minimal failing test case?

**Comment by mrob95** [comment]
For me on py3 the following:
```
from dragonfly import *

test_list = List("test")

class TestRule(MappingRule):
    mapping = {
        "<ref>": Text("%(ref)s")
    }
    extras = [ListRef("ref", test_list)]

grammar = Grammar("test_grammar")
grammar.add_rule(TestRule())
grammar.load()
```

Produces:
```
Traceback (most recent call last):
  File "c:\users\mike\documents\github\dragonfly2\dragonfly\loader.py", line 60, in load
    namespace)
  File "C:\Users\Mike\Documents\GitHub\kaldi_test2\_test.py", line 13, in <module>
    grammar.load()
  File "c:\users\mike\documents\github\dragonfly2\dragonfly\grammar\grammar_base.py", line 35
    self._engine.load_grammar(self)
  File "c:\users\mike\documents\github\dragonfly2\dragonfly\engines\base\engine.py", line 138
    wrapper = self._load_grammar(grammar)
  File "c:\users\mike\documents\github\dragonfly2\dragonfly\engines\backend_kaldi\engine.py",    kaldi_rule_by_rule_dict = self._compiler.compile_grammar(grammar, self)
  File "c:\users\mike\documents\github\dragonfly2\dragonfly\engines\backend_kaldi\compiler.pymmar
    self._compile_rule_root(rule, grammar, kaldi_rule)
  File "c:\users\mike\documents\github\dragonfly2\dragonfly\engines\backend_kaldi\compiler.pyle_root
    kaldi_rule.compile(lazy=self.lazy_compilation)
  File "C:\Users\Mike\AppData\Local\Programs\Python\Python37\lib\site-packages\kaldi_active_grammar\compiler.py", line 92, in compile
    return self.finish_compile()
  File "C:\Users\Mike\AppData\Local\Programs\Python\Python37\lib\site-packages\kaldi_active_grammar\compiler.py", line 99, in finish_compile
    self.compiler._compile_agf_graph(compile=True, nonterm=self.nonterm, input_data=self._fst_text, filename=self.filepath)
  File "C:\Users\Mike\AppData\Local\Programs\Python\Python37\lib\site-packages\kaldi_active_grammar\compiler.py", line 283, in _compile_agf_graph
    compile_command()
  File "C:\Users\Mike\AppData\Local\Programs\Python\Python37\lib\site-packages\ush.py", line 585, in __call__
    return wait(procs, raise_on_error)
  File "C:\Users\Mike\AppData\Local\Programs\Python\Python37\lib\site-packages\ush.py", line 189, in wait
    result = tuple(iterate_outputs(procs, raise_on_error, status_codes))
  File "C:\Users\Mike\AppData\Local\Programs\Python\Python37\lib\site-packages\ush.py", line 221, in iterate_outputs
    raise ProcessError(process_info)
ush.ProcessError: One or more commands failed: [(['C:\\Users\\Mike\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\kaldi_active_grammar\\exec\\windows\\fstcompile', '--isymbols=kaldi_model_zamia\\words.txt', '--osymbols=kaldi_model_zamia\\words.txt'], 11528, 0), (['C:\\Users\\Mike\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\kaldi_active_grammar\\exec\\windows\\compile-graph-agf', '--arcsort-grammar', '--grammar-prepend-nonterm=999995', '--grammar-append-nonterm=999996', '--nonterm-phones-offset=993', '--read-disambig-syms=kaldi_model_zamia\\disambig.int', '--verbose=0', 'kaldi_model_zamia\\tree', 'kaldi_model_zamia\\final.mdl', 'kaldi_model_zamia\\L_disambig.fst', '-', 'kaldi_model_zamia.tmp\\10171f75033001e8dda80fae7dc4692d26093a91.fst'], 15928, 4294967295)]
```

A list like this which has elements in it already doesn't produce an error:
```
test_list = List("test", ["list item"])
```

**Comment by daanzu** [comment]
Ah, I see the problem now. I'll put in a fix.

Regarding problem 2 above, do you think it is resolved by https://github.com/dictation-toolbox/dragonfly/pull/157 ?

**Comment by mrob95** [comment]
I'm not yet sure, I'm going to do some more testing tomorrow

**Comment by daanzu** [comment]
@mrob95 Did https://github.com/dictation-toolbox/dragonfly/pull/159 fix problem 2 for you?

**Comment by mrob95** [comment]
Apologies for the late reply. Yes, I've been using it pretty much full time for the past week and it's been smooth, including with lazy compilation. Thanks!

**Comment by daanzu** [comment]
Great! No problem. I am working on fixing the other problem, which is actually a problem for any empty grammar, but it is a bit tricky.

**Comment by daanzu** [comment]
Problem 1 should now be resolved by upgrading to the just released KaldiAG v1.0.2. I think this fixes the rest of this issue?

**Comment by mrob95** [comment]
Yep, that fixes everything. Thanks again!
